### PR TITLE
CPM-726: Create Domain Objects

### PR DESCRIPTION
**Pull request by antoinearrive**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

**Review by fchoquet — COMMENTED** [review]
Could you maybe add a quick description of each class' responsibility?

Some are obvious, but some are not (for instance "Structure" and "Target' are very generic terms)

Or maybe there is a design document that you can link?

File: components/identifier-generator/back/src/Domain/Model/IdentifierGenerator.php [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class IdentifierGenerator
+{
+    public function __construct(
+        private IdentifierGeneratorCode $identifierGeneratorCode,
+        private Conditions $conditions,
+        private Structure $structure,
+        private LabelCollection $labelCollection,
+        private ?Delimiter $delimiter,
```
Comment (antoinearrive) [inline components/identifier-generator/back/src/Domain/Model/IdentifierGenerator.php]: add at the end

File: components/identifier-generator/back/src/Domain/Model/LabelCollection.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+declare(strict_types=1);
+
+namespace Akeneo\Pim\Automation\IdentifierGenerator\Domain\Model;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class LabelCollection
```
Comment (antoinearrive) [inline components/identifier-generator/back/src/Domain/Model/LabelCollection.php]: final

File: components/identifier-generator/back/src/Domain/Model/Property/AutoNumber.php [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+namespace Akeneo\Pim\Automation\IdentifierGenerator\Domain\Model\Property;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class AutoNumber implements PropertyInterface
+{
+    public function __construct(
+        private int $minimalNumber,
```
Comment (antoinearrive) [inline components/identifier-generator/back/src/Domain/Model/Property/AutoNumber.php:14]: naming ?

File: components/identifier-generator/back/tests/Specification/Domain/Model/TargetSpec.php [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        $this->shouldBeAnInstanceOf(Target::class);
+    }
+
+    function it_cannot_be_instantiated_with_an_empty_string()
+    {
+        $this->beConstructedThrough('fromString', ['']);
+        $this->shouldThrow(\InvalidArgumentException::class)->duringInstantiation();
+    }
+
+    function it_represents_a_target()
```
Comment (cemorin) [inline components/identifier-generator/back/tests/Specification/Domain/Model/TargetSpec.php:32]:
```suggestion
    function it_returns_a_target()
```

File: components/identifier-generator/back/src/Domain/Model/Property/AutoNumber.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ */
+final class AutoNumber implements PropertyInterface
+{
+    public function __construct(
+        private int $numberMin,
+        private int $digitsMin,
+    ) {
+    }
+
+    public static function fromValues(int $numberMin, $digitsMin): self
```
Comment (pierallard) [inline components/identifier-generator/back/src/Domain/Model/Property/AutoNumber.php:21]:
```suggestion
    public static function fromValues(int $numberMin, int $digitsMin): self
```

File: components/identifier-generator/back/src/Domain/Model/Property/FreeText.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+final class FreeText implements PropertyInterface
+{
+    private function __construct(
+        private string $value,
+    ) {
+    }
+
+    public static function fromString(string $value)
+    {
+        $value = trim($value);
```
Comment (pierallard) [inline components/identifier-generator/back/src/Domain/Model/Property/FreeText.php:22]: Remove the trim

File: components/identifier-generator/back/tests/Specification/Domain/Model/Condition/ConditionsSpec.php [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+use PhpSpec\ObjectBehavior;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ConditionsSpec extends ObjectBehavior
+{
+    function it_cannot_be_instantiate_with_not_conditions_interface()
```
Comment (pierallard) [inline components/identifier-generator/back/tests/Specification/Domain/Model/Condition/ConditionsSpec.php:15]: it_only_accepts_condition
```suggestion
    function it_cannot_be_instantiated_with_something_else_than_acondition()
```

File: components/identifier-generator/back/tests/Specification/Domain/Model/IdentifierGeneratorSpec.php [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            $delimiter,
+        );
+    }
+
+    function it_is_a_identifier_generator()
+    {
+        $this->shouldBeAnInstanceOf(IdentifierGenerator::class);
+    }
+
+    function it_can_instantiate_without_delimiter()
```
Comment (pierallard) [inline components/identifier-generator/back/tests/Specification/Domain/Model/IdentifierGeneratorSpec.php:50]:
```suggestion
    function it_can_instantiated_without_delimiter()
```

**Review by pierallard — APPROVED** [review]

File: components/identifier-generator/back/tests/Specification/Domain/Model/Condition/ConditionsSpec.php [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+use PhpSpec\ObjectBehavior;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ConditionsSpec extends ObjectBehavior
+{
+    function it_cannot_be_instantiated_with_something_else_than_acondition()
```
Comment (cemorin) [inline components/identifier-generator/back/tests/Specification/Domain/Model/Condition/ConditionsSpec.php:15]:
```suggestion
    function it_cannot_be_instantiated_with_something_else_than_a_condition()
```

**Review by mmetayer — APPROVED** [review]

**Pull request merged**